File: leetcode/Ruby/No11.container_with_most_water.rb
=begin
Difficulty:
Medium

Desc:
Given n non-negative integers a1, a2, ..., an , where each represents a point at coordinate (i, ai). n vertical lines are drawn such that the two endpoints of line i is at (i, ai) and (i, 0).
Find two lines, which together with x-axis forms a container, such that the container contains the most water.

Example:
Input: [1,8,6,2,5,4,8,3,7]
Output: 49
=end

# @param {Integer[]} height
# @return {Integer}
def max_area(height)
  i = 0
  j = height.size - 1
  max = 0
  while i < j do
    hi = height[i]
    hj = height[j]
    max = [max, (i - j).abs * [hi, hj].min].max
    if hi < hj then
      i += 1
    else
      j -= 1
    end
  end
  return max
end

File: leetcode/C#/No11.container-with-most-water.cs
/*
 * Difficulty:
 * Medium
 *
 * Desc:
 * Given n non-negative integers a1, a2, ..., an , where each represents a point at coordinate (i, ai).
 * n vertical lines are drawn such that the two endpoints of line i is at (i, ai) and (i, 0).
 * Find two lines, which together with x-axis forms a container, such that the container contains the most water.
 * Note: You may not slant the container and n is at least 2.
 * The above vertical lines are represented by array [1,8,6,2,5,4,8,3,7].
 * In this case, the max area of water (blue section) the container can contain is 49.
 *
 * Example:
 * Input: [1,8,6,2,5,4,8,3,7]
 * Output: 49
*/

public class Solution {
    public int MaxArea(int[] height) {
        int i = 0;
        int j = height.Length - 1;
        int res = 0;

        while (i < j) {
            res = Math.Max(
                res,
                Math.Min(height[i], height[j]) * (j - i)
            );
            if (height[i] <= height[j]) {
                i += 1;
            } else {
                j -= 1;
            }
        }
        return res;
    }
}